warehouse/ingest-core/src/test/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregatorTest.java

```diff
@@ -224,8 +225,353 @@ public void testCount() throws Exception {
         Uid.List resultList = Uid.List.parseFrom(result.get());
         
         assertEquals(5, resultList.getCOUNT());
-        assertEquals(false, resultList.getIGNORE());
+        assertFalse(resultList.getIGNORE());
+        assertEquals(1, resultList.getUIDCount());
+        
+    }
+    
```

Comment:
Thanks for tackling this worthwhile effort to correct the inaccuracies with the aggregator!

The test below adds the same UID twice, highlighting the possible inconsistencies between getCount and the size of getUIDList.

Note that the test fails both before and after this PR's changes.  It's just highlighting another inaccuracy with the aggregator.

```
    @Test
    public void testAddUIDTwice() throws Exception {
        GlobalIndexUidAggregator localAgg = new GlobalIndexUidAggregator();
        IteratorSetting is = new IteratorSetting(19, "test", GlobalIndexUidAggregator.class);
        GlobalIndexUidAggregator.setCombineAllColumns(is, true);
        localAgg.validateOptions(is.getOptions());

        UUID uuid1 = UUID.randomUUID();
        ArrayList<Value> values = Lists.newArrayList();
        Builder b = createNewUidList();
        b.setIGNORE(false);
        b.setCOUNT(1);
        b.addUID(uuid1.toString());
        values.add(new Value(b.build().toByteArray()));

        b = createNewUidList();
        b.setIGNORE(false);
        b.setCOUNT(1);
        b.addUID(uuid1.toString());
        values.add(new Value(b.build().toByteArray()));

        Value result = localAgg.reduce(new Key("key"), values.iterator());
        Uid.List resultList = Uid.List.parseFrom(result.get());        
        assertEquals(1, resultList.getUIDCount());
        assertTrue(resultList.getUIDList().contains(uuid1.toString()));
        assertEquals(1, resultList.getUIDList().size());
        assertEquals(1, resultList.getCOUNT()); // Fails here getCount()==2
    }
```

There are use cases in which the same UID will be added multiple times so it's worth considering the behavior.  

The test above highlights how the aggregator will have an inconsistency between its count and list size.  The test below shows how the aggregator will assess that the max was reached without first checking if the UID is unique or already in the UIDList.  

It seems that the same UID added three times (while IGNORE == false) should result in a count of one and a UID list of size one.

```
    @Test
    public void testAddUIDThrice() throws Exception {
// lowered max to show problem more easily
        GlobalIndexUidAggregator localAgg = new GlobalIndexUidAggregator(2);
        IteratorSetting is = new IteratorSetting(19, "test", GlobalIndexUidAggregator.class);
        GlobalIndexUidAggregator.setCombineAllColumns(is, true);
        localAgg.validateOptions(is.getOptions());

        UUID uuid1 = UUID.randomUUID();
        ArrayList<Value> values = Lists.newArrayList();
        Builder b = createNewUidList();
        b.setIGNORE(false);
        b.setCOUNT(1);
        b.addUID(uuid1.toString());
        values.add(new Value(b.build().toByteArray()));

        b = createNewUidList();
        b.setIGNORE(false);
        b.setCOUNT(1);
        b.addUID(uuid1.toString());
        values.add(new Value(b.build().toByteArray()));

        b = createNewUidList();
        b.setIGNORE(false);
        b.setCOUNT(1);
        b.addUID(uuid1.toString());
        values.add(new Value(b.build().toByteArray()));

        Value result = localAgg.reduce(new Key("key"), values.iterator());
        Uid.List resultList = Uid.List.parseFrom(result.get());
        assertEquals(1, resultList.getUIDCount()); // Fails getUIDCount == 0
        assertTrue(resultList.getUIDList().contains(uuid1.toString()));
        assertEquals(1, resultList.getUIDList().size());
        assertEquals(1, resultList.getCOUNT());
    }
```

Perhaps the logic should not depend on count unless seenIgnore is true.  The size of the UID list would be a more accurate count to use, when available.

Note that before the UIDs are put into the protobuf List they are deduplicated in a HashSet within the aggregator.

Maybe this issue is outside the scope of this ticket.

Reply:
Sure, these are good points. I originally wanted only to correct the UIDs getting dropped, and was trying to change as little as possible to make that happen. Then more things cropped up and that lead to adding the released UID list and also re-working the inner logic. It's not unreasonable to try to solve issues with duplicate UIDs too. It gets tricky if you are trying to combine 2 protocol buffers that are near their limit (and have a large max UID count). Do you combine the two sets and have a possibly huge (albeit temporary list)? Do you compute an intersection look at the total count (which may save memory but you also have to do it every time and have the performance hit)?

Reply:
@matthpeterson it turns out there's actually a test that verifies that duplicate UIDs are counted! See [here](https://github.com/NationalSecurityAgency/datawave/blob/integration/warehouse/ingest-core/src/test/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregatorTest.java#L226).

Reply:
@matthpeterson and @jschmidt10, I have pushed changes to attempt to address duplicate UIDs (and also get rid of the quarantine/release lists). It's still a tricky class to get the behavior right, but hopefully this is an improvement. Have a look (especially what's going on with counts in the code and tests) and let me know what you think.

Reply:
Interesting test.  I see now that a lot of the other tests also assert a discrepancy between getCount and getUIDCount().

If a system ever re-ingests the same data and then removes it, the existing logic will leave "dangling" index entries - even for a list that has never hit the max.  I'd rather the logic be changed to be as accurate as possible.  I don't know why we wanted to deliberately have inconsistencies between the count and UIDList sizes except to accommodate the quarantine / release logic.  Now that the logic is being updated to rely on the two separate lists for quarantine and release, I don't think we need to try to conflate the meaning of count.

> It gets tricky if you are trying to combine 2 protocol buffers that are near their limit (and have a large max UID count). Do you combine the two sets and have a possibly huge (albeit temporary list)? Do you compute an intersection look at the total count (which may save memory but you also have to do it every time and have the performance hit)?

The change below seems to correct the discrepancies between getCount() and getUIDCount() and make the two unit tests from my prior comment pass.
```
--- a/warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java
+++ b/warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java
@@ -87,9 +87,9 @@ public class GlobalIndexUidAggregator extends PropogatingCombiner {

         // as a backup, we remove the intersection of the UID sets

-        builder.setCOUNT(count);
+        builder.setCOUNT(seenIgnore ? count : uids.size());

-        if (seenIgnore || count > maxUids) {
+        if (seenIgnore || uids.size() > maxUids) {
```

This change would require making the getCOUNT and getUIDCount assertions match for the test you mentioned and a few others.
```
@@ -224,13 +224,81 @@ public class GlobalIndexUidAggregatorTest {
         Value result = agg.reduce(new Key("key"), values.iterator());
         Uid.List resultList = Uid.List.parseFrom(result.get());

-        assertEquals(5, resultList.getCOUNT());
+        assertEquals(1, resultList.getCOUNT());
```
and
```
@@ -267,7 +335,7 @@ public class GlobalIndexUidAggregatorTest {
         Collections.reverse(values);
         Value result = localAgg.reduce(new Key("key"), values.iterator());
         Uid.List resultList = Uid.List.parseFrom(result.get());
-        assertEquals(1, resultList.getCOUNT());
+        assertEquals(2, resultList.getCOUNT());
         assertEquals(2, resultList.getUIDCount());
         assertEquals(2, resultList.getUIDList().size());
         assertEquals(0, resultList.getREMOVEDUIDList().size());
@@ -318,7 +386,7 @@ public class GlobalIndexUidAggregatorTest {
         Collections.reverse(values);
         Value result = localAgg.reduce(new Key("key"), values.iterator());
         Uid.List resultList = Uid.List.parseFrom(result.get());
-        assertEquals(-2, resultList.getCOUNT());
+        assertEquals(1, resultList.getCOUNT());
         assertEquals(1, resultList.getUIDCount());
         assertEquals(1, resultList.getUIDList().size());
         assertEquals(3, resultList.getREMOVEDUIDCount());
@@ -476,7 +544,7 @@ public class GlobalIndexUidAggregatorTest {
         Collections.reverse(values);
         Value result = localAgg.reduce(new Key("key"), values.iterator());
         Uid.List resultList = Uid.List.parseFrom(result.get());
-        assertEquals(0, resultList.getCOUNT());
+        assertEquals(1, resultList.getCOUNT());
         assertEquals(1, resultList.getUIDCount());
         assertEquals(1, resultList.getUIDList().size());
         assertEquals(1, resultList.getREMOVEDUIDList().size());
@@ -543,7 +611,7 @@ public class GlobalIndexUidAggregatorTest {
         Collections.reverse(values);
         Value result = localAgg.reduce(new Key("key"), values.iterator());
         Uid.List resultList = Uid.List.parseFrom(result.get());
-        assertEquals(1, resultList.getCOUNT());
+        assertEquals(2, resultList.getCOUNT());
```

Reply:
> The change below seems to correct the discrepancies between getCount() and getUIDCount() and make the two unit tests from my prior comment pass.

I had already made a similar change in my latest commit. I think I must have pushed right as you were typing this up @matthpeterson. Have a look at the latest. It's a bit more of a refactoring since the quarantine list was also removed. It might be easier at this point to look at the raw file vs the diffs.

Reply:
I should add that I looked at enforcing the constraint that `getCOUNT() == (getUIDCount() - getREMOVEDUIDCount())` but found this to not be possible. When we go through the UID list to add them to the in-memory set, we first check if the UID is in the uidsToRemove set and take no action if that's the case. One might suggest removing the uid from uidsToRemove at that point since we've encountered the "add" that goes with the remove. However, it would not be safe to remove the uid from uidsToRemove since we don't yet know if there are any other adds for that uid (from other entries that we haven't seen yet).